src/main/scala/zio/interop/javaconcurrent.scala

```diff
@@ -25,85 +26,155 @@ import scala.concurrent.ExecutionException
+      Task.effectAsync { k =>
+        val handler = new CompletionHandler[T, Any] {
+
+          def completed(result: T, u: Any): Unit = k(Task.succeed(result))
+
+          def failed(t: Throwable, u: Any): Unit = t match {
+            case e if !p.fatal(e) => k(Task.fail(e))
+            case _                => k(Task.die(t))
           }
-          cb(io)
         }
-      }
 
-    def fromCompletionStage[A, E >: Throwable](csIo: IO[E, CompletionStage[A]]): IO[E, A] =
-      csIo.flatMap(unsafeCompletionStageToIO)
-
-    def fromCompletionStage[A](cs: () => CompletionStage[A]): Task[A] =
-      IO.effectSuspendTotal {
-        unsafeCompletionStageToIO(cs())
+        try {
+          op(handler)
+        } catch {
+          case e if !p.fatal(e) => k(Task.fail(e))
+        }
       }
+    }
 
-    private def unsafeFutureJavaToIO[A](future: Future[A]): Task[A] = {
-      def unwrap[B](f: Future[B]): Task[B] =
-        IO.flatten {
-          IO.effectTotal {
-            try {
-              val result = f.get()
-              IO.succeed(result)
-            } catch {
-              case e: ExecutionException =>
-                IO.fail(e.getCause)
-              case _: InterruptedException =>
-                IO.interrupt
-              case t: Throwable => // CancellationException
-                IO.fail(t)
+  private def catchFromGet(isFatal: Throwable => Boolean): PartialFunction[Throwable, Task[Nothing]] = {
+    case e: CompletionException =>
+      Task.fail(e.getCause)
+    case e: ExecutionException =>
+      Task.fail(e.getCause)
+    case _: InterruptedException =>
+      Task.interrupt
+    case e if !isFatal(e) =>
+      Task.fail(e)
+  }
+
+  private def unwrapDone[A](isFatal: Throwable => Boolean)(f: Future[A]): Task[A] =
+    try {
+      Task.succeed(f.get())
+    } catch catchFromGet(isFatal)
+
+  def fromCompletionStage[A](csUio: UIO[CompletionStage[A]]): Task[A] =
+    csUio.flatMap { cs =>
+      Task.effectSuspendTotalWith { p =>
+        val cf = cs.toCompletableFuture
+        if (cf.isDone) {
+          unwrapDone(p.fatal)(cf)
+        } else {
+          Task.effectAsync { cb =>
+            val _ = cs.handle[Unit] { (v: A, t: Throwable) =>
+              val io = Option(t).fold[Task[A]](Task.succeed(v)) { t =>
+                catchFromGet(p.fatal).lift(t).getOrElse(Task.die(t))
+              }
+              cb(io)
             }
           }
         }
+      }
+    }
 
-      if (future.isDone) {
-        unwrap(future)
-      } else {
-        blocking(unwrap(future)).provide(Blocking.Live)
+  /** WARNING: this uses the blocking Future#get, consider using `fromCompletionStage` */
+  def fromFutureJava[A](futureUio: UIO[Future[A]]): Task[A] =
+    futureUio.flatMap { future =>
+      Task.effectSuspendTotalWith { p =>
+        if (future.isDone) {
+          unwrapDone(p.fatal)(future)
+        } else {
+          blocking(Task.effectSuspend(unwrapDone(p.fatal)(future))).provide(Blocking.Live)
```

Comment:
@sideeffffect Should we hardcode the pool or leave the choice to the user via `RIO[Blocking, A]`? The signature requiring `Blocking` would also serve as a warning about the underlying blocking

Reply:
:heavy_check_mark: thanks for the tip